Cookbook/Proposals/MealyReactiveFeedback.md

```diff
@@ -0,0 +1,141 @@
+
+```swift
+let incrementFeedback = Feedback<Int, Event> { _ in // NOTE: argument is not used
+    incrementSignal.map { _ in Event.increment }
+}
+```
+
+Notice that `Signal<State>` argument is not used at all.
+While this code works in general, mixing these 2 different semantics causes verbosity and sometimes confusion.
+
+The ideal approach will be to **separate "external events" as plain `Signal<Event>`, and pass it as a new argument next to `Feedback`s**.
+
+The idea of "external event" has been previously discussed in [Babylonpartners/ReactiveFeedback#32](https://github.com/Babylonpartners/ReactiveFeedback/pull/32), and it also makes sense from automata theory that "external events" (alphabets) and "feedbacks" (output function) are different.
+Also, Elm provides [`Subscription`](https://package.elm-lang.org/packages/elm/core/latest/Platform-Sub) which works just as same as observing "external events" and it is different from feedback type called `Cmd<Event>`.
+
+Furthermore, by separating "external event" nicely, we can send "initial external event" only after system is instantiated, so that the racing issue described in [ReactiveFeedback/pull#38](https://github.com/Babylonpartners/ReactiveFeedback/pull/38#issuecomment-468478325) will no longer occur.
+
+### 3. Remove duplicated `state` caching in both `SignalProducer` and `Property`
+
+`State` is stored in both `scan` and `Property` which is obviously inefficient.
+
+## Proposed solution
+
+### Changes
+
+```diff
+ public struct Feedback<State, Event> {
+-    let events: (Scheduler, Signal<State, NoError>) -> Signal<Event, NoError>
++    let events: (Scheduler, Signal<(Event, State), NoError>) -> Signal<Event, NoError>
+ }
+
+ extension Property {
+     public convenience init<Event>(
+         initial: Value,
+         scheduler: Scheduler = QueueScheduler.main,
+         reduce: @escaping (Value, Event) -> Value,
++        events: Signal<Event>,   // external events
+         feedbacks: [Feedback<Value, Event>]
+     ) {
+         ...
+     }
+ }
+```
+
+- `Feedback` preserves the streaming format, i.e. `Signal<???> -> Signal<Event>` rather than `??? -> Signal<Event>` so that it can have flattening capability e.g. `flatMap(.latest)` or `flatMap(.first)`.
+- Resolves [Babylonpartners/ReactiveFeedback#32](https://github.com/Babylonpartners/ReactiveFeedback/pull/32)
+- Improves [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44)
+    - Resolves "idle -> routing -> idle" navigation state transition dance.
+- **Backward compatible** with current Moore machine (we can still add as many states for precise control as we want)
+
+### Usage
+
+```swift
+let externalEvent = Signal.merge(
+    increment.map { Event.increment },
+    decrement.map { Event.decrement }
+)
+
+let system = Property(
+    initial: 0,
+    reduce: counterReducer,
+    events: externalEvent,
```

Comment:
This is good 👍  Now we want need to use `Feedback.input`

Reply:
I'd personally prefer `input`, or `eventInput` as a name for this parameter, to disambiguate it with events that are being produced inside the system itself.

I can see some similarities with disturbance (`Gd(s)`) in the control theory, if we assume the disturbance is the signal of events which is summed up with a signal of events in the system itself (Gc(s))

<img width="1284" alt="Screenshot 2019-04-22 at 17 10 09" src="https://user-images.githubusercontent.com/1488293/56510379-874ff080-6521-11e9-93cf-c693f9b8b39f.png">
https://ocw.mit.edu/courses/mechanical-engineering/2-017j-design-of-electromechanical-robotic-systems-fall-2009/lecture-notes/MIT2_017JF09_control.pdf

Reply:
@ilyapuchka https://github.com/Babylonpartners/ios-playbook/pull/98#discussion_r277337817

I'm not familiar with control theory (which I believe is an interesting topic with regard to FRP and RAF), but as far as glancing its basics, I think:

- `R(s)` is the input (we are currently calling `externalEvent`)
- `H(s)` is the feedback
- `Gd(s)` is a noise (we probably don't need to care)

In contrast to FRP, external events and feedback events can't be combined together with `+`/`-`.

Reply:
I'm also +1 with `input` over `events`, so I will update the code :)